neo/io/nwbio.py

```diff
@@ -0,0 +1,811 @@
+    unit_name = prefix_map[conversion] + base_unit_name
+    if unit_name[-1] == "s":  # strip trailing 's', e.g. "volts" --> "volt"
+        unit_name = unit_name[:-1]
+    return getattr(pq, unit_name)
+
+
+class NWBIO(BaseIO):
+    """
+    Class for "reading" experimental data from a .nwb file, and "writing" a .nwb file from Neo
+    """
+    supported_objects = [Block, Segment, AnalogSignal, IrregularlySampledSignal,
+                         SpikeTrain, Epoch, Event, ImageSequence]
+    readable_objects = supported_objects
+    writeable_objects = supported_objects
+
+    has_header = False
+    support_lazy = True
+
+    name = 'NeoNWB IO'
+    description = 'This IO reads/writes experimental data from/to an .nwb dataset'
+    extensions = ['nwb']
+    mode = 'one-file'
+
+    is_readable = True
+    is_writable = True
+    is_streameable = False
+
+    def __init__(self, filename, mode='r'):
+        """
+        Arguments:
+            filename : the filename
+        """
+        if not have_pynwb:
+            raise Exception("Please install the pynwb package to use NWBIO")
+        if not have_hdmf:
+            raise Exception("Please install the hdmf package to use NWBIO")
+        BaseIO.__init__(self, filename=filename)
+        self.filename = filename
+        self.blocks_written = 0
+        self.nwb_file_mode = mode
+
+    def read_all_blocks(self, lazy=False, **kwargs):
+        """
+
+        """
+        assert self.nwb_file_mode in ('r',)
+        io = pynwb.NWBHDF5IO(self.filename, mode=self.nwb_file_mode)  # Open a file with NWBHDF5IO
+        self._file = io.read()
+
+        self.global_block_metadata = {}
+        for annotation_name in GLOBAL_ANNOTATIONS:
+            value = getattr(self._file, annotation_name, None)
+            if value is not None:
+                if annotation_name in POSSIBLE_JSON_FIELDS:
+                    value = try_json_field(value)
+                self.global_block_metadata[annotation_name] = value
+        if "session_description" in self.global_block_metadata:
+            self.global_block_metadata["description"] = self.global_block_metadata["session_description"]
+        self.global_block_metadata["file_origin"] = self.filename
+        if "session_start_time" in self.global_block_metadata:
+            self.global_block_metadata["rec_datetime"] = self.global_block_metadata["session_start_time"]
+        if "file_create_date" in self.global_block_metadata:
+            self.global_block_metadata["file_datetime"] = self.global_block_metadata["file_create_date"]
+
+        self._blocks = {}
+        self._read_acquisition_group(lazy=lazy)
+        self._read_stimulus_group(lazy)
+        self._read_units(lazy=lazy)
+        self._read_epochs_group(lazy)
+
+        return list(self._blocks.values())
+
+    def read_block(self, lazy=False, **kargs):
+        """
+        Load the first block in the file.
+        """
+        return self.read_all_blocks(lazy=lazy)[0]
```

Comment:
This implementation is quite strange.
generaly, the read_all_blocks() loop over read_block()

I also would add a block_index here instead of always use 0